apis/goharbor.io/v1alpha2/chartmuseum_types.go

```diff
@@ -0,0 +1,356 @@
+
+	// +kubebuilder:validation:Required
+	Chart ChartMuseumChartSpec `json:"chart"`
+}
+
+type ChartMuseumServerSpec struct {
+	// +kubebuilder:validation:Optional
+	TLS *harbormetav1.ComponentsTLSSpec `json:"tls,omitempty"`
+
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:validation:Type="string"
+	// +kubebuilder:validation:Pattern="([0-9]+h)?([0-9]+m)?([0-9]+s)?([0-9]+ms)?([0-9]+us)?([0-9]+µs)?([0-9]+ns)?"
+	// Socket timeout
+	ReadTimeout *metav1.Duration `json:"readTimeout,omitempty"`
+
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:validation:Type="string"
+	// +kubebuilder:validation:Pattern="([0-9]+h)?([0-9]+m)?([0-9]+s)?([0-9]+ms)?([0-9]+us)?([0-9]+µs)?([0-9]+ns)?"
+	// Socket timeout
+	WriteTimeout *metav1.Duration `json:"writeTimeout,omitempty"`
+
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:validation:Minimum=0
+	// +kubebuilder:default=20971520
+	// Max size of post body (in bytes)
+	MaxUploadSize *int64 `json:"maxUploadSize,omitempty"`
+
+	// +kubebuilder:validation:Optional
+	// Value to set in the Access-Control-Allow-Origin HTTP header
+	CORSAllowOrigin string `json:"corsAllowOrigin,omitempty"`
+}
+
+type ChartMuseumChartSpec struct {
+	// +kubebuilder:validation:Optional
+	// Form fields which will be queried
+	PostFormFieldName ChartMuseumPostFormFieldNameSpec `json:"postFormFieldName,omitempty"`
+
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Pattern="https?://.*"
+	// The absolute url for .tgzs in index.yaml
+	URL string `json:"url"`
+
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default=true
+	// Allow chart versions to be re-uploaded without ?force querystring
+	AllowOvewrite *bool `json:"allowOverwrite,omitempty"`
+
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default=false
+	// Enforce the chart museum server only accepts the valid chart version as Helm does
+	SemanticVersioning2Only bool `json:"onlySemver2"`
+
+	// +kubebuilder:validation:Required
+	Storage ChartMuseumChartStorageSpec `json:"storage"`
+
+	// +kubebuilder:validation:Optional
+	Index ChartMuseumChartIndexSpec `json:"index,omitempty"`
+
+	// +kubebuilder:validation:Optional
+	Repo ChartMuseumChartRepoSpec `json:"repo,omitempty"`
+}
+
+type ChartMuseumChartRepoSpec struct {
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default=false
+	// The length of repo variable
+	DepthDynamic bool `json:"depthDynamic"`
+
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:validation:Minimum=0
+	// +kubebuilder:default=1
+	// Levels of nested repos for multitenancy
+	Depth *int32 `json:"depth,omitempty"`
+}
+
+type ChartMuseumChartStorageSpec struct {
+	ChartMuseumChartStorageDriverSpec `json:",inline"`
+
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:validation:Minimum=0
+	// Maximum number of objects allowed in storage (per tenant)
+	MaxStorageObjects *int64 `json:"maxStorageObject,omitempty"`
+}
+
+type ChartMuseumChartStorageDriverSpec struct {
```

Comment:
Here we only support the three below storage driver. What about others? Can they be suppirted by overriding the registry configurations directly?

```
inmemory: A temporary storage driver using a local inmemory map. This exists solely for reference and testing.
filesystem: A local storage driver configured to use a directory tree in the local filesystem.
s3: A driver storing objects in an Amazon Simple Storage Service (S3) bucket.
azure: A driver storing objects in Microsoft Azure Blob Storage.
swift: A driver storing objects in Openstack Swift.
oss: A driver storing objects in Aliyun OSS.
gcs: A driver storing objects in a Google Cloud Storage bucket.
```

Reply:
Registry configuration can no more be savagely overridden. This is limiting but safer.

`azure`, `oss` and `gcs` are totally strangers for me.
I do not event know where to start to have credentials. :s
If anyone is used to these platform, please help.
Otherwise I will do that later...

Do not worry, I am adding this limitation to the readme.
And it should not be that long to develop these drivers.